tests/src/drivers/signature.c

```diff
@@ -0,0 +1,299 @@
+/*
+ * Test driver for signature functions
+ */
+/*  Copyright The Mbed TLS Contributors
+ *  SPDX-License-Identifier: Apache-2.0
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License"); you may
+ *  not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ *  WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+#if !defined(MBEDTLS_CONFIG_FILE)
+#include "mbedtls/config.h"
+#else
+#include MBEDTLS_CONFIG_FILE
+#endif
+
+#if defined(MBEDTLS_PSA_CRYPTO_DRIVERS) && defined(PSA_CRYPTO_DRIVER_TEST)
+#include "psa/crypto.h"
+#include "mbedtls/ecp.h"
+
+#include "drivers/signature.h"
+
+#include "mbedtls/md.h"
+#include "mbedtls/ecdsa.h"
+
+#include "test/random.h"
+
+#include <string.h>
+
+/* If non-null, on success, copy this to the output. */
+void *test_driver_forced_output = NULL;
+size_t test_driver_forced_output_length = 0;
+
+psa_status_t test_transparent_signature_sign_hash_status = PSA_ERROR_NOT_SUPPORTED;
+unsigned long test_transparent_signature_sign_hash_hit = 0;
+
+psa_status_t test_transparent_signature_verify_hash_status = PSA_ERROR_NOT_SUPPORTED;
+unsigned long test_transparent_signature_verify_hash_hit = 0;
+
+psa_status_t test_transparent_signature_sign_hash(
+    const psa_key_attributes_t *attributes,
+    const uint8_t *key, size_t key_length,
+    psa_algorithm_t alg,
+    const uint8_t *hash, size_t hash_length,
+    uint8_t *signature, size_t signature_size, size_t *signature_length )
+{
+    ++test_transparent_signature_sign_hash_hit;
+
+    if( test_transparent_signature_sign_hash_status != PSA_SUCCESS )
+        return( test_transparent_signature_sign_hash_status );
+
+    if( test_driver_forced_output != NULL )
+    {
+        if( test_driver_forced_output_length > signature_size )
+            return( PSA_ERROR_BUFFER_TOO_SMALL );
+        memcpy( signature, test_driver_forced_output,
+                test_driver_forced_output_length );
+        *signature_length = test_driver_forced_output_length;
+        return( PSA_SUCCESS );
+    }
+
+    psa_status_t status = PSA_ERROR_NOT_SUPPORTED;
+
+#if defined(MBEDTLS_ECDSA_C) && defined(MBEDTLS_ECDSA_DETERMINISTIC) && \
```

Comment:
My understanding of the code below is that it is a light version of `psa_sign_hash()` (didn't check completely though, but if it isn't I am not sure to understand why anyway).
The high-level structure of `psa_sign_hash()` is:
`psa_get_key_from_slot();`
`psa_driver_wrapper_sign_hash();`
`if not done do the operation with software`
Thus what about to reorganize in psa_crypto.c the "do the operation with software" in one function and just call this function here.
That way there should be less code duplication, thus less code maintenance and it makes clear that the code here is just the code in psa_crypto.c.

Reply:
That would mean exposing that "do the operation with software" function through a header, such that the test driver can reach it. I've had pushback before on this, since it can lead to confusion as to what the public API is.

Not saying I don't want to do it, but then it should be done consequently for all "do in software" operations.

Reply:
We have a [process](https://github.com/ARMmbed/mbedtls/blob/development/docs/architecture/testing/invasive-testing.md#user-content-reliance-on-internal-details) for this: declare the internal function in a `library/*.h`, and if there's no other reason to make it non-`static`, use `MBEDTLS_STATIC_TESTABLE` instead of `static`.

Reply:
Yes it should be done consistently and the "do the operation with software" should be in an internal header thus currently in an header in library (not in include/psa).

Reply:
Since this will lead to further splintering of the psa_crypto.c core, can we delegate this work to a separate work unit?

Reply:
I would be good to get the whole picture for the PSA API as stake in this PR thus for generate_key(), sign_hash() and verify_hash(). Thus I would be in favor of doing this work in this PR. When we are ok with the overall result for those APIs we would generalized this in a separate PR.

Reply:
> This process is fairly new so there aren't many examples in the code base yet.
That's an understatement: There's exactly *one*. And it came in after I started this PR. Meaning *yet another* rebase. I'm going to forego the static declaration part for now and just declare the internal functions in an internal header. We previously established that the test drivers shouldn't be dependant on `MBEDTLS_TEST_HOOKS`, so the internal functions not being accessible unless that is defined is in issue regardless.

Reply:
I'm personally ok with duplicating some code in the test drivers. It's code that uses stable interfaces so it's unlikely to break much, and we don't care all that much if there's a bug fix in the copied code that doesn't get applied to the copy in the test code. As long as these are clearly test drivers (as opposed to sample drivers), it's ok if they don't have the best hygiene and it's ok if they're buggy or even insecure in corner cases.

Reply:
The _internal functions that got split off when implementing driver delegation are now made `MBEDTLS_STATIC_TESTABLE`, which would make them available for basing an eventual fully-featured test driver off of them. Creating that fully featured test driver is however not in scope for this PR.

Reply:
OK, changed to unconditional external linkage. This may bloat code size for those not using LTO, though.

Reply:
> please remove your last commit

Only the last commit? I.e. you want to keep the `MBEDTLS_STATIC_TESTABLE` declarations? Or do you want me to revert the whole splitting up of the functions such that the software fallback is inlined in the `psa_crypto_xxx_hash()` again?

Reply:
I wrote this before you pushed the latest patch (following Gilles comment). Thus I propose to remove (rather than revert) the two last commits from this PR.

Reply:
Done.